ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs

```diff
@@ -0,0 +1,744 @@
+{-# LANGUAGE TypeFamilies         #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -Wno-orphans -Wredundant-constraints #-}
+module Test.Ouroboros.Storage.ChainDB.StateMachine where
+
+import           Prelude hiding (elem)
+
+import           Codec.Serialise (decode, encode)
+import           Control.Monad (void)
+import           Data.Bifunctor (first)
+import           Data.Foldable (toList)
+import           Data.Functor.Classes (Eq1, Show1)
+import qualified Data.Map as Map
+import           Data.Proxy
+import           Data.TreeDiff (ToExpr)
+import           Data.Typeable
+import           Data.Word (Word64)
+import           GHC.Generics (Generic, Generic1)
+
+import qualified Generics.SOP as SOP
+
+import           Test.QuickCheck
+import qualified Test.QuickCheck.Monadic as QC
+import           Test.StateMachine
+import qualified Test.StateMachine.Types as QSM
+import qualified Test.StateMachine.Types.Rank2 as Rank2
+import           Test.Tasty (TestTree, testGroup)
+import           Test.Tasty.QuickCheck (testProperty)
+
+import           Control.Monad.Class.MonadSTM
+import           Control.Monad.Class.MonadThrow
+
+import           Ouroboros.Network.AnchoredFragment (AnchoredFragment)
+import           Ouroboros.Network.Block (BlockNo (..), ChainHash (..),
+                     HasHeader, HeaderHash, Point, SlotNo (..), StandardHash)
+import qualified Ouroboros.Network.Block as Block
+import           Ouroboros.Network.Chain (Chain (..), chainToList, genesisPoint)
+import           Ouroboros.Network.ChainProducerState (ChainProducerState,
+                     ReaderNext, ReaderState)
+import qualified Ouroboros.Network.ChainProducerState as CPS
+
+import           Ouroboros.Consensus.Ledger.Abstract
+import           Ouroboros.Consensus.Protocol.Abstract
+
+import           Ouroboros.Storage.ChainDB hiding (ChainUpdate (..))
+import qualified Ouroboros.Storage.ChainDB.Model as Model
+import qualified Ouroboros.Storage.FS.Sim.MockFS as Mock
+import           Ouroboros.Storage.FS.Sim.STM (simHasFS)
+import           Ouroboros.Storage.ImmutableDB
+                     (ValidationPolicy (ValidateAllEpochs))
+import           Ouroboros.Storage.LedgerDB.MemPolicy (defaultMemPolicy)
+import qualified Ouroboros.Storage.Util.ErrorHandling as EH
+
+import           Test.Ouroboros.Storage.ChainDB.TestBlock
+
+import           Test.Util.RefEnv (RefEnv)
+import qualified Test.Util.RefEnv as RE
+
+
+{-------------------------------------------------------------------------------
+  Abstract model
+-------------------------------------------------------------------------------}
+
+-- | Commands
+data Cmd blk hdr it
+  = AddBlock         blk
+  | GetCurrentChain
+  | GetCurrentLedger
+  | GetTipBlock
+  | GetTipHeader
+  | GetTipPoint
+  | GetBlock         (Point blk)
+  | StreamBlocks     (StreamFrom blk) (StreamTo blk)
+  | IteratorNext     it
+  | IteratorClose    it
+  deriving (Generic, Show, Functor, Foldable, Traversable)
```

Comment:
We need to test corruption too. Important: including when using iterators and readers.
Some call to the ImmutableDB, the VolatileDB, or the LedgerDB encountered corruption: see what happens and whether the other DBs are properly synched/recovered/whatever.

Don't try to be too clever, because we might miss some tings: just simulate errors using the existing error simulation infrastructure. The underlying database will notice and the recovery will be triggered.
But don't simulate things like: a block was not written to disk, but then a GC happened afterwards, etc.

We can add a command that triggers garbage collection: this can hopefully trigger the race conditions in the iterators that have to deal with blocks moving between databases. The GC command will have no parameters, it will just perform the scheduled GC. Also see the last section in `ChainDB.md`.

Reply:
See #653.